File: TODO.py
'''
# TODO(Noah):
- No need for forward decl of functions.
- add the alloc keyword as an expression (kind of like var assignment/func decl).
- add the free keyword as a statement.
- >> operator for piping function calls to object instances
    - I'll have it in the syntax parser as an expression, because '>>' is an operator.

# TODO(Noah): To add many more operators.
- The attribute of a variable . (also put implicit derefencing in there)
- function call on a variable ()
- indexing on a variable []
- +=, -=, and this whole family
# https://en.cppreference.com/w/c/language/operator_precedence

TODO(Noah): Check out the things below,
- Consider branching and Phi nodes for compiler optimization.
- Check out what vectorizin is.
- Look into quantization (making ops smaller size)?
- What is a GPR?
'''

'''
TODO(Noah):
We want to get the unix console target working. Along with linking with a standard library.
What do we need to do to get this example working?

Terminal things
- ppl::get_attached_term()
- ppl::terminal
- ppl::terminal::get_attr()
- ppl::terminal_attr
- ppl::terminal::set_atr(ppl::terminal_attr)
- ppl::terminal::set_default_attr()

Other standard libs
- ppl::seed_rand()
- ppl::time()
- ppl::rand()
- ppl::system(string str);
- ppl::flush()
- ppl::block_until_fready(ppl::STDIN, 0)
- ppl::getchar()
- ppl::sleep(100000);

Language Features:
- const modifier on types.
- initializer lists
    Ex) const []char FOOD_SYMBOLS = {'%', '$', '&', '@', '+'};
- dynamic arrays
    Ex) [dynamic]snake_body snake;
    snake.append(body);
- accessor operator
    Ex) body.x
- ++ op
    Ex) for (int i(0); i < 5; i++)
- [] operator
    Ex) snake[0].x
- len operator on arrays (static or dyanmic)
    Ex) for (int i = 1; i < len(snake); i++) {
- bool datatype
- modulus operator 
    Ex) ppl::rand() % (MAP_WIDTH - 3);
- sizeof operator
    Ex) ppl::rand() % sizeof(FOOD_SYMBOLS);
- double variable initialiazation
    Ex) int last_x, last_y;
- replacement of default with case:
    Ex) case: break; // NOTE(Noah): Same as default:
- += op
    Ex) SCORE += 7;
- bitwise OR, &= (bitwise and equals), ~ (whatever this unary op is)
    Ex) attr.local_modes &= ~( ICANON | ECHO );
'''

'''TODO
- Need to implement proper targeting of platforms.
    - Right now it is really difficult to look forward and be like: 'Yes, this is how
    we mght implement the other platforms and therefore we should prepare by architecting
    our code like ___.
    Not going to work. Let's just keep it in mind that this is the end goal. In the meantime
    everything is to target the console target.
'''

'''
# NOTE(Noah): utf8 things.
string we recieve as input from the user are based on the locale set in the system.
anything we pass to printf is also supposed to be in this encoding.
wide characters are NOT utf8. They are consistent width, 32 bit per character encodings.
wprintf and %ls do these wide character things.
we want multibyte encoding.
'''


'''
NOTE: Things I like from the Odin programming language.
https://odin-lang.org/

Note that I do not like the entire language, and hence the reason for a new one.
Like, it's not digestible. It is TOO different from C...
Need something easily digested.
Odin is literally backwards.
like...
    a : int 
that shit is backwards.

TODO: List of things I like that we should implement.

    - 0b01010101 (binary literals)
    - 0xFF (Hexadecimal literals)
    - Complex numbers, 2i + 3
    - 1.0e9 (scientific notation)
    - 1_000_000_000 (underscores for better readability)
    - 64 bit floating point numbers
    - a, b = 3 (multi-variable assignment)

    - removed while loops and make them a for loop...

    (range-based for-loop)
    - for int j in 0..<10 
        // Certain built-in types can be iterated over
        some_string := "Hello, 世界"
        for character in some_string { // Strings are assumed to be UTF-8
            fmt.println(character)
        }
        for value in some_array {
            fmt.println(value)
        }

    - (indices in range-based for-loops)
        for char character, int index in some_string {
            fmt.println(index, character)
        }

    - (range-based for-loop, iterate by reference)
        for ->char character, int index in some_string {
            fmt.println(index, character)
        }

    - when statement. Constant expression evaluated at compile time. I see this as a replacement
    for #ifdef except we can do expressions as well.

        when ODIN_ARCH == "386" {
			fmt.println("32 bit")
		} else when ODIN_ARCH == "amd64" {
			fmt.println("64 bit")
		} else {
			fmt.println("Unsupported architecture")
		}
    
    - implicit dereference has the benefit of when you need to refactor code to use
    a pointer versus a non-pointer, you don't have to go around and do all these
    '.' vs '->' op changes.
    # NOTE(Noah): Still need to think about this. It's GOOD to know that you are working w/ a pointer.

    - struct literal
    v = Vector3{1, 4, 9}

    - make Quaternion a default type.

    - # NOTE(Noah): Consider polymorphism.

    - functions defined inside the scope of a another function definition.
        - NOTE(Noah): This might be one of my favourite features :)
    
    - lambda functions.
        - NOTE(Noah): I also really like this one. Not sure if from Odin, but from Python.

    - list initializers 
        i.e. []int myArray = {3,4,5,6,7};

    - add a map type (like a hash map)
        - NOTE(Noah): I love these :)

    - list slicing
        - NOTE(Noah): Fun :) One of my personal favs from Python

    - big numbers and precision mathematics.

    - put vectors and matrices in the language.
        - kind of like numpy. Make it good for that type of shit.
'''

File: TODO.cpp
// C operator precedence, for reference
// # https://en.cppreference.com/w/c/language/operator_precedence

// NOTE(Noah): Apparently everything good about language design has already been done...
// So what if we made PPL flexible? -> metaprogramming?

// NOTE(Noah): Currently considering range based for-loops.
    // because there are common patterns where I want to change the index as it iterates
    // and I feel that this is odd when using a range based for-loop.

// NOTE(Noah): Currently considering implicit dereference.
    // because if you have this, then you might not know that you are dealing with a pointer.
    // this is not explicit, and I generally dislike a lack of explicitness.
    // NOTE(Noah): But what if there was some other feedback mechanism?
        // Like it syntax highlighted your pointer variable?    

/* TODO

NOTE(Noah): There are complexities with the package system that will only become apparent once we start to implement it.
Thus, we stop with what we have defined here and simply get on with our day. 
Desired Package System design:
Gonna use the keyword, "import".
- You have a file, and at the top you specify what your code is
dependent on. That's it. Every package is namespaced.
    -> It might be a header file that defines some routines. Maybe you will be linking against object files
    in the build step?
    -> A single header file library ( aka another .ppl file or a .h/.c/.cpp file ).
        -> This is to get at the single-file compilation model
        -> These are automatically include guarded
- Implications of this package system
    - Can import the main file into sub-components to leverage global variables that might be stored in the 
    main namespace.
- Each namespace defines orderless, callable methods and ordless variables with storage.
    - global variables are initialized in an orderless manner and deinit in orderless manner.
    - but this happens at the start of the program and then they all are destructed at the end of the program.

Roadmap:

    - Port Compiler codebase to C/C++.
        - Port syntax.py for AST tree generation and add appropriate tests in tests.cpp
            - this is a parse grammer test.
        - Add a preparser that does nothing but implementing a package system (port preparser.py).
        - Port codegen.py and linker.py to complete the port to C
            - While we are doing this, ensure the C backend is actually a C backend, not a C++ backend.
            - But this begs the question, which version of C? C89? C99? Modern C?
            - Here is how I am going to make my decision.
                - This is a MODERN language, and as such, we are going to use MODERN C.
                - Simple as that.
                - Because we can expect compilers to support this.
            - AND because the C backend is going to become irrelevant (likely) as soon as we implement the LLVM
            IR backend.
            - linker.py is to be ported not to C but to a batch script that takes the compiler output (C backend) then runs additional
            tooling to get it down to the executable.

    - Is the language in a usable state, and are we satisfied with the language?
        - Look at the C89 spec and your own projects to answer this question.
        - C89 spec: http://port70.net/~nsz/c/c89/c89-draft.html#2.

    - Things to add to language to get into "usable state" / to feel "satisfied".
        - Verify that names of functions are legal (cannot begin with $ for example)
        - Can define structs/functions anywhere. Like: inside another function if wanted.
        - 0b01010101 (binary literals)
        - 0xFF (Hexadecimal literals)
        - 1.0e9 (scientific notation)
        - default values in structs.
        - default parameters in functions OVER function overloading (that's a design choice).
            -> Reason being is explicitness.
        - struct literal
            v = Vector3{1, 4, 9}
        - Add all desired variable types and sizes Ex) 
            - dynamic arrays
                - negative indices work (Python got this right).
            - maps
            - string (a utf8 string).
                - you can add strings together.
                - you can concat characters to strings.
                - you can check if two strings are equal.
                - you can index into strings.
                - string literals.
            - char (a unicode coidepoint). 
            - uint64 (unsigned 64 bit integer).
            - float (32 bit floating point).
            - double (64 bit floating point).
            - vector
            - matrix
        - Ensure all operators are in the language
            - linear algebra ops with vectors and matrices.
            - add the len() operator for getting the length of
                - strings
                - dynamic arrays
                - static array
            - Add bitshift operators. Ex) >> and <<.
            - Ensure all postfix and prefix are there Ex) ++X or X++
            - Add -> for dereffing pointers. OR decide on implicit dereferencing.
        - Ensure standard library has what it needs
            - Single character printing of unicode code points.

    - Add LLVM IR as backend
        - In the meantime, ignore the standard library.

    - Put in debugging symbols so that the binary can be stepped through while also
    viewing the source code at the same time.
        - these symbols are placed in object files, which have a symbol table.
        - I actually do not have any idea how this might look, and will only know until I dive into the weeds
        of the LLVM compiler project...can't wait! :')

    - Right now, the standard library for PPL is a layer on top of the standard C library.
        - This is not a portable solution.
        - As such, this is how the standard library is going to work.
            - It has a different version for every target architecture.
            - We make as little platforms specific parts of the library, making the standard library
            modular.

    - Go ahead and write some targets!
        - Implement a rock-solid target for POSIX compliant systems.
        - Write in a Web target.
        - Write in an IOS target.
        - Write an Anroid target.
        - Write in a Windows Console target.
        - QR codes :P
        Desktop GUI/Game targets
            - Write in a Windows GUI target
                - Need to think about different graphics APIs 
            - Write in a macOS GUI target (Metal).

    - Once we write a web target.
        - Build a PPL website so that people can learn how to download and use the project
        for themselves.
        - Make sure to add analytics to the website.

    - Make PPL easy to download and use for different host architectures.
        - POSIX System (macOS, any flavour of Linux, etc)
        - Windows

    - Implement inline assembly
        - THEN Write a quick and dirty x86 assembly output so that we can build oeprating systems with PPL.
            - Good thing about this is that there will be no need to implement a standard library. 
        - Then port Falcon-OS to be compiled via PPL.
            - And ensure this process is EASY -> so maybe I have to think about Virtual Box/Qemu!

    - Add some features to the standard library.
        - Such as Wifi/Bluetooth protocols.
        - Multithreading.
        
    - AirDeploy

    - OKAY. So the LLVM IR is yet a stepping stone.
        - Problem: Was going to do optimization before LLVM IR step. but seems LLVM IR optimizes even after this when going down to machine
        code.
        - For the purpose of learning, and potentially doing it better, we will try to "beat LLVM IR -> machine code optimization".
        - Or maybe see that we can create an optimized LLVM IR ourselves, such that the LLVM IR has "no work" to do when it does its own
        optimization step.
            - Because this way I can reap the benefits of portability that the LLVM IR offers.     
    
    - Add support for Hot Reloading

    - Add support for Hot Debugging


Other Tasks (Can be done anytime):
- add the alloc keyword as an expression (kind of like var assignment/func decl).
- add the free keyword as a statement.
- >> operator for piping function calls to object instances
    - I'll have it in the syntax parser as an expression, because '>>' is an operator.
- double variable initialiazation
    Ex) int last_x, last_y;
- "stringification" of enums.
- lambda functions.
- Complex numbers, 2i + 3
- 1_000_000_000 (underscores for better readability)
- make Quaternions a default type.

Check out the things below,
- Consider branching and Phi nodes for compiler optimization.
- Check out what vectorizin is.
- Look into quantization (making ops smaller size)?
- What is a GPR?
*/